From 6dbc0923f4f2365b3cdbc7f1061ae8adc6a65e03 Mon Sep 17 00:00:00 2001
From: Mousumi Jana <mousumix.jana@intel.com>
Date: Fri, 11 Dec 2015 00:01:50 +0530
Subject: [PATCH 001/102] ASoC: Intel: Support for loading the machine driver
 using module parameter

Added support for loading the machine driver for RSE and HU using
module parameter.For loading the RSE or HU machine driver
"bxt_ivi_rse_i2s" or "bxt_ivi_hu_i2s" parameter should be
passed as a module parameter.

Change-Id: Icc948a9dfdf24d9b1f305fb0f9d784e077566102
Signed-off-by: Mousumi Jana <mousumix.jana@intel.com>
Reviewed-on: https://git-gar-1.devtools.intel.com/gerrit/7818
Reviewed-by: Babu, Ramesh <ramesh.babu@intel.com>
Tested-by: Babu, Ramesh <ramesh.babu@intel.com>

Conflicts:
	sound/soc/intel/skylake/skl.c
---
 sound/soc/intel/skylake/skl.c | 36 ++++++++++++++++++++++++++++++------
 1 file changed, 30 insertions(+), 6 deletions(-)

diff --git a/sound/soc/intel/skylake/skl.c b/sound/soc/intel/skylake/skl.c
index 674dd5c..d8f92f6 100644
--- a/sound/soc/intel/skylake/skl.c
+++ b/sound/soc/intel/skylake/skl.c
@@ -39,7 +39,19 @@
 #include "skl-sst-ipc.h"
 #include "skl-topology.h"
 
+static struct skl_machine_pdata skl_dmic_data;
+
+static char *machine = NULL;
+static char *fw_filename = NULL;
+
+module_param(machine, charp, 0444);
+MODULE_PARM_DESC(machine, "machine driver string for Intel soundcard.");
+
+module_param(fw_filename, charp, 0444);
+MODULE_PARM_DESC(fw_filename, "firmware filename string for Intel DSP.");
+
 /*
+
  * initialize the PCI registers
  */
 static void skl_update_pci_byte(struct pci_dev *pci, unsigned int reg,
@@ -515,14 +527,24 @@ static int skl_find_machine(struct skl *skl, void *driver_data)
 	    IS_ENABLED(CONFIG_SND_SOC_INTEL_CNL_FPGA))
 		goto out;
 
-	mach = snd_soc_acpi_find_machine(mach);
-	if (mach == NULL) {
-		dev_err(bus->dev, "No matching machine driver found\n");
-		return -ENODEV;
+	if (machine) {
+		mach = kzalloc(sizeof(*mach), GFP_KERNEL);
+		if (!mach)
+			return -ENOMEM;
+		mach->drv_name = machine;
+		if (fw_filename)
+			mach->fw_filename = fw_filename;
+	} else {
+		mach = driver_data;
+		mach = snd_soc_acpi_find_machine(mach);
+		if (mach == NULL) {
+			dev_err(bus->dev, "No matching machine driver found\n");
+			return -ENODEV;
+		}
+
 	}
 
 out:
-
 	skl->fw_name = mach->fw_filename;
 	skl->mach = mach;
 	if (mach->pdata) {
@@ -539,7 +561,6 @@ static int skl_machine_device_register(struct skl *skl)
 	struct snd_soc_acpi_mach *mach = skl->mach;
 	struct platform_device *pdev;
 	int ret;
-
 	pdev = platform_device_alloc(mach->drv_name, -1);
 	if (pdev == NULL) {
 		dev_err(bus->dev, "platform device alloc failed\n");
@@ -558,6 +579,9 @@ static int skl_machine_device_register(struct skl *skl)
 
 	skl->i2s_dev = pdev;
 
+	if (machine)
+		kfree(mach);
+
 	return 0;
 }
 
-- 
1.9.1

